### DISCUSSION TITLE
Need more details before publishing extension

### DISCUSSION THREAD

**Discussion by msorens** [microsoft/vscode-discussions#504] [Extension Development QnA] [answered]
The [documentation](https://code.visualstudio.com/api/working-with-extensions/publishing-extension#publishing-extensions) on publishing is insufficient for managing product integrity and release. Let me explain. 
To publish an extension the docs say to either run `vsce publish` or upload from the VSMarketplace management page, the latter merely having an "upload file" button.

I need to know what happens next.
After the file is uploaded is it immediately made public? I am guessing not, but I cannot tell my company management "I _think_ so." There is a screenshot on the doc page that shows that there exists an "unpublished" state, which implies there is a setup step then a go live step, but again, I cannot go to my management with such "maybe"s.

My specific questions:
1. From the command line, if I use `vsce publish` (with or without `--pre-release`), does the extension go public immediately?
2. Same question if I upload on the management web page; does it go public immediately?
3. Is there a way to NOT go public immediately? (May have already been answered by (1) or (2).

**Comment by gjsjohnmurray** [comment]
In my experience, paths 1 and 2 produce the same effect. The extension doesn't publish **immediately** but it usually only takes a few minutes for Marketplace verification checks to complete, after which it is available for anyone to install.

Re 3, I'm not aware of a way to upload an extension but keep it in a "pending publication" state that I can subsequently release it from.

**Comment by alefragnani** [comment, marked as answer]
Hi @msorens,

For questions 1 and 2, the simplest answer is **no**. The extension won't go public immediately. At every new publish you do, the extension will be _validated_ by the marketplace, which takes a few minutes, most of the time. I remember at least these 3 steps that will happen at marketplace level when you publish an extension:

* checks if your extension has more than 10 keywords (https://github.com/microsoft/vscode-discussions/discussions/426)
* sign the extension (https://github.com/microsoft/vscode-discussions/discussions/137) and a possible publisher sign in the future
* checks for viruses

> another set of validations are made by `vsce` itself by the way, like supported badges, presence of svgs in readme, etc

For question 3, on the other hand, the answer is **not yet, but maybe in the future**. The marketplace team is working on an _unlisted extension_ concept (https://github.com/microsoft/vsmarketplace/issues/292). At first, it won't support unlisted _versions_ but the extension as a whole, but the discussion is still open. I suggest you to take a closer look and contribute in the discussion, to show your scenario for the team.

Hope this helps

**Reply by msorens** [reply]
Thanks for the reference on the unlisted extension! That would have been ideal for my current needs of releasing a NEW extension so that I can make it available for only specific users to easily install. I say "would have been" only because of time: I need it now and there is no timeframe even suggested for that enhancement. Sigh...

**Comment by msorens** [comment]
One thing still puzzles me, seemingly at odds with the two replies thusfar: this illustration from the documentation clearly shows an "unpublished" state. Where does that come from?
<img width="753" alt="image" src="https://user-images.githubusercontent.com/6817500/222871190-a2761ace-cb30-4a48-bef7-f293d336b82b.png">

**Reply by alefragnani** [reply]
Not aware of, unfortunately. 

But, since your question is very specific about the marketplace, I suggest you to crosspost it at their repo (https://github.com/microsoft/vsmarketplace) too. They are a different team, and could probably answer it easier.

**Reply by alefragnani** [reply]
In fact, it must be _unpublish extension_, as described here https://code.visualstudio.com/api/working-with-extensions/publishing-extension#unpublishing-extensions

**Reply by msorens** [reply]
Ah, quite so. So that is unpublish _after_ it has already been published, not starting with unpublished.... (Hence the need for the unlisted option, as you indicated.)

Thanks again!

**Comment by PEZ** [comment]
To me it is quite clear. We ask for the extension to get published, wether using vsce or the web panel, and it gets published. Given that it passes the checks.
